README.md
## tp-xfan (ThinkPad XForms Fan)

`tp-xfan` is a simple program written in C which lets users with
any ThinkPad (that has ACPI fan support) modify their fan's speed,
going from 0 (auto), to a maximum of 8 (disengaged).

It also features an intuitive yet nostalgic Graphical User Interface,
written using XForms (hence the name, XForms Fan).

For more features, check [Features](#Features).

For a list of changes and/or news, check the `NEWS`
and `ChangeLog` files.

Initially written on May 28, 2021.
Current version can be found in `tp-xfan.c` under `src/`,
or by running `make version` (see [Installation](#Installation))

If you plan to work on `tp-xfan`, be sure to read `INDENTATION`.

Support this project: https://saloniamatteo.top/donate.html

## Flags
Currently, `tp-xfan` supports the following command-line flags:

| Flag | Argument | Description                                      |
|------|----------|--------------------------------------------------|
| `-d` | None     | Enable Debugging                                 |
| `-i` | None     | Print program info and exit (No XForms required) |
| `-h` | None     | Print help and exit                              |
| `-p` | Required | Use a custom path for the apply.sh script        |
| `-v` | None     | Be more verbose                                  |

## Features
`tp-xfan` has a lot of powerful features, including:

+ Value Slider to choose values
+ Setting a custom directory for the apply.sh script,
in case you don't have it in your current directory,
nor under `/usr/local/share/tp-xfan`.
+ AFSM (Automatic Fan Speed Management); See [AFSM](#AFSM).
+ Automatic recognition of root & non-root users
	- When running as root, no password is required,
	and the command is executed immediately, if the
	apply.sh script was found.
	- When running as a non-root user, the root
	password is asked (but not stored anywhere),
	in order to access `/proc/acpi/ibm/fan`.
	- When running as a non-root user with
	a setuid program (like sudo/doas), no
	password is required from `tp-xfan`.
+ Local Password Caching: the password is stored in a 
variable, which will only be accessible by `tp-xfan` itself.
+ Debugging functionalities, to see what `tp-xfan` really is doing
even when you cannot see it.
+ Verbosity, to see if the command to set the fan speed
actually worked or not.

Note that Debugging and Verbosity only output to the console.

## AFSM
Automatic Fan Speed Management (or AFSM) is a very powerful
feature of `tp-xfan`. It allows users to choose a CPU temperature
threshold, the speed of the fan, and the interval (in seconds).

For example, with a given threshold temperature of 50°C, AFSM will
start the fan after the seconds provided by the interval,
with the speed chosen by the user.

So, every interval seconds, AFSM will check if the current CPU
temperature is above the threshold: if it is, it will
keep going, setting the fan to the speed chosen by the user;
if the temperature is below, or equal to the threshold,
it will immediately set the fan speed to 0 (auto),
and it will cycle forever.

## Screenshots
Start Menu: the initial screen, shown when `tp-xfan` is opened.

![Start Menu](https://raw.githubusercontent.com/saloniamatteo/tp-xfan/master/pics/1-init-scr.png)

Info Screen: showing info.

![Info Screen](https://raw.githubusercontent.com/saloniamatteo/tp-xfan/master/pics/2-info-scr.png)

Password Screen: `tp-xfan` is asking the root password.

![Password Screen](https://raw.githubusercontent.com/saloniamatteo/tp-xfan/master/pics/3-pass-scr.png)

AFSM Screen: [Auto Fan Speed Management](#AFSM) screen.

![AFSM Screen](https://raw.githubusercontent.com/saloniamatteo/tp-xfan/master/pics/4-afsm-scr.png)

Exit Confirm: `tp-xfan` is asking the user if they really want to exit.

![Exit Confirm](https://raw.githubusercontent.com/saloniamatteo/tp-xfan/master/pics/5-exit-confirm.png)

## Requirements
In order to run `tp-xfan`, you need the following packages:

+ `xforms` (**required**): provides the core Forms functionality.
(Gentoo: x11-libs/xforms, others: http://xforms-toolkit.org)
+ `expect` (<u>optional</u>): used for escalating privileges,
when not running as root. (Gentoo: dev-tcltk/expect)
+ Linux's `su` (<u>optional</u>): used for escalating privileges,
when not running as root. Guaranteed to be present on a
fully working Linux system.

Also, you need to have the `thinkpad_acpi` module.
If you build your Kernel yourself, make sure the following are enabled:

```
Device Drivers --->
	[*] X86 Platform Specific Device Drivers  --->  
		<M>   ThinkPad ACPI Laptop Extras          
```

Then, make sure the driver is available:

`lsmod | grep thinkpad_acpi`

Then, write the following to `/etc/modprobe.d/thinkpad.conf`:

```
options thinkpad_acpi fan_control=1
```

Now, you should reboot. After rebooting, you should be able to
modify your ThinkPad's fan speed.

## Installation
(See [Requirements](#Requirements))

`tp-xfan` uses GNU AutoTools to increase its portabilty and flexibility.

Normally, users should run the following commands, to install `tp-xfan`:

```bash
./configure
make
# the following should be ran as root,
# or by prepending "sudo" or "doas".
make install install-data
```

This will install `tp-xfan`, and the default `apply.sh` script (/usr/local/share/tp-xfan/apply.sh)

If, for some reason, you cannot run the commands above, run `autoreconf --install`, then retry.

## Help
If you need help, you can:
- Create an issue;
- Open a pull request;
- Send me an email [(saloniamatteo@pm.me](mailto:saloniamatteo@pm.me) or [matteo@mail.saloniamatteo.top)](mailto:matteo@mail.saloniamatteo.top).


NEWS
News/ChangeLog for tp-xfan (ThinkPad XForms Fan).
============================================================

[Release 1.2] 06/06/2021:
	Support multi-threading using POSIX threads
	(pthreads). This allows users to use AFSM
	in a non-blocking way, after pressing "Apply & Run"


src/temp.h
/* See LICENSE file for copyright and license details.
 *
 * This header, temp.h, provides the function get_temp,
 * to get the current CPU temperature.
 *
 * Made by Salonia Matteo <saloniamatteo@pm.me>
 *
 */

#ifndef TEMP_H
#define TEMP_H

#include <stdio.h>
#include <stdlib.h>

/* Function prototypes */
int get_temp(void);

int
get_temp(void)
{
	FILE *fd;
	char temp_str[6], ch;
	int temp = 0, i = 0;

	fd = fopen("/sys/class/thermal/thermal_zone0/temp", "r");

	/* Cannot read temperature */
	if (fd == NULL) {
		fprintf(stderr, "Cannot read temperature! Exiting.\n");
		return -1;
	}

	/* Read temperature */
	ch = fgetc(fd);

	while (ch != EOF) {
		temp_str[i] = ch;
		i++;
		ch = fgetc(fd);
	}

	temp_str[i - 1] = '\0';

	/* Convert string to integer */
	temp = atoi(temp_str) / 1000;

	/* Close file descriptor */
	fclose(fd);

	return temp;
}

#endif	/* TEMP_H */


src/apply.sh
#!/bin/sh
# This script sets the level/speed of the ThinkPad fan.
# Must be run as root.
# Written by Salonia Matteo on 28/05/2021.

# check if we are root
if [ $EUID != 0 ]; then
	printf "You must be root to run this! Exiting with ENOTRECOVERABLE.\n"
	exit 131 # ENOTRECOVERABLE
fi

# Check if user passed an argument
if [ -z "$1" ]; then
	printf "Please enter a value! Exiting with EILSEQ.\n"
	exit 84 # EILSEQ
else
	level="$1"
fi

# Check speed level
if [[ $level = 0 || $level = "auto" ]]; then
	level="auto"
elif [[ $level = 8 || $level = "max" ]]; then
	level="disengaged"
elif [[ $level < 0 || $level > 8 ]]; then
	printf "Speed level not valid! Exiting with EOVERFLOW.\n"
	exit 75 # EOVERFLOW
fi

while [ $(cat /proc/acpi/ibm/fan | grep level -m1 | awk '{print $2}') != "$level" ]; do
	echo "level $level" | $root tee "/proc/acpi/ibm/fan"
done

exit 0 # Success


src/tp-xfan.c
/* See LICENSE file for copyright and license details.
 *
 * tp-xfan (ThinkPad XForms Fan) is a program which
 * uses the XForms C library to display in a beautiful
 * way some controls, in order to modify the speed of
 * the ThinkPad's fan. It supports automatic speed
 * (level 0), all the way up to maximum speed (level 8).
 *
 * It also provides AFSM, or Auto Fan Speed Management,
 * which lets users modify in a powerful way
 * how should the fan run, etc.
 *
 * Made by Salonia Matteo <saloniamatteo@pm.me>
 *
 */

#define _GNU_SOURCE
#define _FORTIFY_SOURCE 2
#define _POSIX_C_SOURCE 200809L

#ifdef HAVE_CONFIG_H
#include "config.h"
#ifndef HAVE_PTHREAD_H
#warning "tp-xfan requires a pthread implementation for AFSM, but found none!"
#endif /* HAVE_PTHREAD_H */
#else
#define PACKAGE_BUGREPORT "saloniamatteo@pm.me"
#define PACKAGE_STRING "tp-xfan 1.2"
#endif /* HAVE_CONFIG_H */

/* For some reason, the following isn't defined in config.h. */
#ifndef PACKAGE_DONATE
#define PACKAGE_DONATE "https://saloniamatteo.top/donate.html"
#endif /* PACKAGE_DONATE */

#include <errno.h>
#include <forms.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>
#include "temp.h"

#define DEFAULT_LOCAL_PATH	"/apply.sh"
#define DEFAULT_SYS_PATH	"/usr/local/share/tp-xfan/apply.sh"
#define THINKPAD_FAN_FILE	"/proc/acpi/ibm/fan"

#define DEBUG(where, ...) if (debug) fprintf(stderr, "Debug [" where "]: " __VA_ARGS__);

#define DEF_TEMP 50	/* Default valslider temperature */
#define MAX_INFO 500	/* Maximum size for thr_info objects */
#define MAX_TEMP 100	/* Maximum valslider temperature */
#define MIN_TEMP 40	/* Minimum valslider temperature */
#define PASS_MAX 100	/* Password maximum length */

struct thr_info {
	FL_OBJECT *obj;
	long data;
} thr_info;

static FL_FORM	 *current_form = {0};	/* Current active form */
static char	 pass[PASS_MAX] = {-1};	/* Cache password, without prompting user */
static char	 path[PATH_MAX] = {0};	/* Path containing apply.sh script */
static char	 user_path[PATH_MAX];	/* User-defined directory that contains the script */
static int	 debug = 0;		/* Enable Debugging */
static int	 fan_rec = 7;		/* Copy of fan value (used to recover AFSM speed) */
static int	 fan_val = 0;		/* Fan speed value */
static int	 interval = 0;		/* AFSM interval (in seconds) */
static int	 is_afsm_on = 0;	/* Is AFSM on? */
static int	 temp = DEF_TEMP;	/* CPU temperature threshold */
static int	 using_custom_path = 0;	/* Are we using a custom directory? */
static int	 verbose = 0;		/* Verbosity */

static char	*INFO_STR = "tp-xfan is a program written in C99 to modify\n\
the speed of the ThinkPad fan, hence the name tp (ThinkPad),\n\
xfan (XForms fan).\n\n\
The only dependencies it has are XForms itself,\n\
the standard C library (which everyone has),\n\
Linux's 'su' command, and TCL's 'expect' command.\n\
\n\
These last two are used to escalate user privileges,\n\
when the user enters the root password. This is necessary,\n\
because we need to send data to /proc/acpi/ibm/fan,\n\
and this requires root privileges.\n\
\n\
Alternatively, you can run this program using\n\
a setuid program, like sudo/doas, or you can run it directly\n\
as root, and it will not ask for a password, since EUID is 0.";

/* Function prototypes */
static void set_valslider(FL_OBJECT *, int, int, int, int, int, void *, long);
static void set_path(void);
static int  check_root(void);
static int  can_write_fan(void);
static void ask_pass(void);
static void showinfo(FL_OBJECT *, long);
static void checkexit(FL_OBJECT *, long);
static void askexit(FL_OBJECT *, long );
static void set_speed(FL_OBJECT *, long);
static void apply_speed(FL_OBJECT *, long);
static void set_afsm_val(FL_OBJECT *, long);
static void set_afsm_secs(FL_OBJECT *, long);
static void afsm(FL_OBJECT *, long);

/* Setup valslider */
static void
set_valslider(FL_OBJECT *obj, int min, int max, int precision, int val, int align, void *func, long data)
{
	fl_set_slider_bounds(obj, min, max);
	fl_set_slider_precision(obj, precision);
	fl_set_slider_value(obj, val);
	fl_set_object_lalign(obj, align);
	fl_set_object_callback(obj, func, data);
	fl_set_slider_return(obj, FL_RETURN_CHANGED);
}

/* Set path */
static void
set_path(void)
{
	/* File descriptor */
	FILE *fd;

	/* If we aren't using a user-defined path,
	 * set path to $PWD/DEFAULT_LOCAL_PATH */
	if (!using_custom_path) {
		DEBUG("Set Path", "Not using custom path.\n");

		/* Get current directory */
		char cwd[PATH_MAX] = { 0 };
		if (getcwd(cwd, sizeof(cwd)) == NULL) {
			fprintf(stderr, "Unable to get current working directory! "
					"Exiting with ENOTRECOVERABLE.\n");
			exit(ENOTRECOVERABLE);
		}

		DEBUG("Set Path", "Gotten current working directory: %s.\n", cwd);

		/* Copy default local path to path */
		snprintf(path, PATH_MAX, "%s%s", cwd, DEFAULT_LOCAL_PATH);

		DEBUG("Set Path", "Set path to %s.\n", path);

	/* Otherwise, set path to $USER_DIR/DEFAULT_LOCAL_PATH */
	} else {
		DEBUG("Set Path", "Using custom directory: %s.\n", user_path);

		if (user_path != NULL)
			snprintf(path, PATH_MAX, "%s%s", user_path, DEFAULT_LOCAL_PATH);
		else {
			fprintf(stderr, "Invalid directory! Exiting with ENOENT.\n");
			exit(ENOENT);
		}

		DEBUG("Set Path", "Set path to %s.\n", path);
	}

	/* Check if apply.sh exists */
	fd = fopen(path, "r");

	/* Does not exist */
	if (fd == NULL) {
		DEBUG("Set Path", "Warning: could not open %s, trying %s.\n",
			path, DEFAULT_SYS_PATH);

		strncpy(path, DEFAULT_SYS_PATH, strlen(DEFAULT_SYS_PATH) + 1);
		fd = fopen(path, "r");

		/* No other files to try, exit. */
		if (fd == NULL) {
			fprintf(stderr, "Error: unable to find %s nor %s!\n",
				DEFAULT_LOCAL_PATH, DEFAULT_SYS_PATH);

			exit(ENOENT);
		}

	}

	/* File was found */
	fclose(fd);
}

/* Check if current user is root */
static int
check_root(void)
{
	uid_t euid = geteuid();

	DEBUG("Check Root", "Gotten EUID: %d\n", euid);

	return (euid == 0 ? 1 : 0);
}

/* Check if we can write to ThinkPad ACPI fan file */
static int
can_write_fan(void)
{
	/* File descriptor */
	FILE *fd;

	fd = fopen(THINKPAD_FAN_FILE, "w");

	if (fd == NULL) {
		DEBUG("Can Write Fan File", "Unable to write to fan file.\n");
		return 0;
	} else {
		DEBUG("Can Write Fan File", "Able to write to fan file.\n");
		fclose(fd);

		return 1;
	}
}

/* Ask password */
static void
ask_pass(void)
{
	/* Check if we have to set the password,
	 * if password wasn't cached */
	if (pass[0] == -1) {
askpass:
		DEBUG("Ask Pass", "Asking root password to user.\n");

		char *pass_l = (char *)fl_show_input("Enter the root password:", "pass");

		if (pass_l == NULL || !strcmp(pass_l, " ") || !strcmp(pass_l, "") || !strcmp(pass_l, "(null)")) {
			fl_show_alert("Error!", "Enter a valid password!", "", 1);
			goto askpass;
		}

		/* Copy local password to password cache */
		DEBUG("Ask Pass", "Copying local password to password cache.\n");
		strncpy(pass, pass_l, PASS_MAX);
		
		DEBUG("Ask Pass", "Gotten password.\n");
	}
}

/* Show program information */
static void
showinfo(FL_OBJECT *obj, long data)
{
	DEBUG("Show Info", "Showing info via XForms.\n");

	fl_show_messages_f("%s written by Salonia Matteo.\n"
			   "Report any bugs to %s\nSupport this project: %s\n\n"
			   "%s",
			   PACKAGE_STRING, PACKAGE_BUGREPORT, PACKAGE_DONATE, INFO_STR);

	DEBUG("Show Info", "Done showing info via XForms.\n");
}

/* Check if user really wants to exit */
static void
checkexit(FL_OBJECT *obj, long data)
{
	// data = 1 -> yes
	if (data) {
		DEBUG("Check Exit", "User wants to exit, destroying Forms.\n");
		fl_finish();
		DEBUG("Check Exit", "Exiting.\n");
		exit(0);
	// data = 0 -> no
	} else {
		DEBUG("Check Exit", "User does NOT want to exit, destroying Exit Form.\n");
		fl_free_form(current_form);
	}
}

/* Exit Confirmation Dialog */
static void
askexit(FL_OBJECT *obj, long data)
{
	DEBUG("Ask Exit", "Building Exit Form via XForms.\n");

	/* Create new XForms objects */
	FL_FORM *exit_form;
	FL_OBJECT *yes, *no;

	DEBUG("Ask Exit", "Adding Objects to Exit Form.\n");

	exit_form = fl_bgn_form(FL_UP_BOX, 320, 120);
	current_form = exit_form;

	/* Add text to form center */
	fl_add_box(FL_UP_BOX, 160, 40, 0, 0, "Do you really want to Quit?");
	
	/* Add buttons and button handlers */
	yes = fl_add_button(FL_NORMAL_BUTTON, 40, 70, 80, 30, "Yes");
	fl_set_object_callback(yes, checkexit, 1);

	no = fl_add_button(FL_NORMAL_BUTTON, 200, 70, 80, 30, "No");
	fl_set_object_callback(no, checkexit, 0);

	current_form = exit_form;

	fl_end_form();

	DEBUG("Ask Exit", "Done Adding Objects to Exit Form.\n");

	/* Show form */
	fl_show_form(exit_form, FL_PLACE_CENTER, FL_TRANSIENT,
			"Exit Confirmation Dialog");

	fl_do_forms();

	DEBUG("Ask Exit", "Showing Exit Form.\n");

	current_form = exit_form;

	fl_free_form(exit_form);

	DEBUG("Ask Exit", "Exit Form destroyed.\n");
}

/* Set fan speed */
static void
set_speed(FL_OBJECT *obj, long data)
{
	DEBUG("Set Speed", "Setting fan speed variable to %d.\n", fan_val);

	fan_val = fl_get_slider_value(obj);

	/* If we're getting called from AFSM,
	 * copy fan_val to fan_rec */
	fan_rec = fan_val;

	DEBUG("Set Speed", "Done Setting fan speed variable.\n");
}

/* Appy speed */
static void
apply_speed(FL_OBJECT *obj, long data)
{
	/* Set path */
	set_path();

	/* If data != 0, set fan_val to data */
	if (data != 0)
		fan_val = data;

	DEBUG("Apply Speed", "Gotten fan value: %d\n", fan_val);

	/* Append speed value to path */
	char val[3] = { 0 };
	snprintf(val, 3, " %d", fan_val);
	strncat(path, val, strlen(path) + 2);

	DEBUG("Apply Speed", "Added speed value [%s] to path: %s.\n", val, path);

	/* Check if current user is root or not */
	int is_root = 0;

	if (can_write_fan() || check_root()) {
		is_root = 1;
		DEBUG("Apply Speed", "User can write to %s, not asking password.\n",
					THINKPAD_FAN_FILE);

	/* User is not root, ask root password */
	} else {
		DEBUG("Apply Speed", "User is NOT root, asking password.\n");
		ask_pass();
	}

	int size = (is_root != 0 ? strlen(path) + 2 : strlen(pass) + strlen(path) + 80);
	char cmd[size];

	DEBUG("Apply Speed", "Set size to %d; is_root: %d\n", size, is_root);

	/* Assemble the command to run */
	if (!is_root) {
		snprintf(cmd, sizeof(cmd),
			"expect -c 'spawn su - root -c \"%s\"; "
			"expect \"Password: \"; send \"%s\n\"; interact;'",
			path, pass);
	} else
		snprintf(cmd, sizeof(cmd), "%s", path);

	DEBUG("Apply Speed", "Set command to run. (For security reasons, it is not printed)\n");

	/* Without verbosity, do not show command output */
	if (verbose == 0)
		strncat(cmd, " > /dev/null", 12);

	DEBUG("Apply Speed", "Running command.\n");

	/* Try to run script */
	int ret = system(cmd);

	DEBUG("Apply Speed", "Command finished.\n");

	if (ret != 0)
		fprintf(stderr, "Unknown error!\n");
}

/* Set min/max values for AFSM */
static void
set_afsm_val(FL_OBJECT *obj, long data)
{
	DEBUG("Set AFSM val", "Setting CPU temperature threshold.\n");
	temp = fl_get_slider_value(obj);
	DEBUG("Set AFSM val", "Set CPU temperature to %d.\n", temp);
}

/* Set AFSM interval */
static void
set_afsm_secs(FL_OBJECT *obj, long data)
{
	DEBUG("Set AFSM seconds", "Setting AMFS Interval.\n");
	interval = fl_get_slider_value(obj);
	DEBUG("Set AFSM seconds", "Set AMFS Interval to %d.\n", interval);
}

/* Run AFSM */
static void
afsm_run(struct thr_info *info)
{
	DEBUG("AFSM Run", "Starting AFSM.\n");

	/* Get obj and data from info */
	FL_OBJECT *obj = info->obj;
	long data = info->data;

	/* While AFSM is active */
	while (is_afsm_on != 0) {
		/* Check CPU temperature after interval seconds */
		fl_msleep(interval * 1000);

		/* Check CPU temperature */
		if (get_temp() < temp) {
			DEBUG("AFSM Run", "Current temp (%d) is less than threshold (%d).\n",
						get_temp(), temp);

			/* Set fan speed to auto */
			DEBUG("AFSM Run", "Setting fan speed to auto.\n");

			fan_val = 0;
			apply_speed(obj, data);

		} else {
			DEBUG("AFSM Run", "Current temp (%d) is greater than (or equal to) threshold (%d).\n",
						get_temp(), temp);

			/* Set & Apply desired speed */
			fan_val = fan_rec;
			DEBUG("AFSM Run", "Fan value: %d; Applying speed.\n", fan_val)
			apply_speed(obj, fan_val);
		}
	}

	/* If we somehow jump out of the loop, reset AFSM status */
	is_afsm_on = 0;
}

/* Stop AFSM */
static void
afsm_stop(struct thr_info *info)
{
	/* Get obj and data from info */
	FL_OBJECT *obj = info->obj;
	long data = info->data;

	/* Reset AFSM status */
	is_afsm_on = 0;

	// data = 1 -> close AFSM form
	if (data == 2) {
		fl_free_form(current_form);
		DEBUG("AFSM Stop", "Destroyed AFSM Form.\n");
	// data = 0 -> stop AFSM
	} else {
		DEBUG("AFSM Stop", "Stopping afsm.\n");

		fan_val = 0;
		apply_speed(obj, data);

		DEBUG("AFSM Stop", "Set fan speed to auto.\n");
	}
}

/* Function to multi-thread AFSM */
static void
afsm_thr(FL_OBJECT *obj, long data)
{
	DEBUG("AFSM Thread", "Starting multi-threaded AFSM.\n");

	/* Variables/objects used for pthreads (only one argument
	 * can be passed when calling pthread_create) */
	pthread_t thr_fan, thr_stp;
	struct thr_info info[MAX_INFO] = { 0 };

	DEBUG("AFSM Thread", "Setting thread info values.\n");

	info->obj = obj;
	info->data = data;

	/* Check data:
	 * 0 -> run			(loop)
	 * 1 -> set fan speed to auto 	(no loop)
	 * 2 -> stop afsm		(no loop) */
	switch (data) {

	/* Run */
	case 0:
		DEBUG("AFSM Thread", "Gotten 0 (run).\n");

		/* Before running, we need to cache the password,
		 * in order to run automatically, so call apply_speed */
		apply_speed(obj, data);

		/* Check if is_afsm_on != 0 */
		if (!is_afsm_on) {
			DEBUG("AFSM Thread", "AFSM not active, starting.\n");

			is_afsm_on++;

			/* Create a thread, which will loop forever,
			 * unless is_afsm_on != 0 */
			pthread_create(&thr_fan, NULL, (void *) afsm_run, info);

		/* Otherwise, do nothing */
		} else
			DEBUG("AFSM Thread", "AFSM is already active!\n");

		break;

	/* Set fan speed to auto, or stop */
	case 1:
	case 2:
		if (data == 1) {
			DEBUG("AFSM Thread", "Gotten 1 (set speed to auto).\n");
		} else
			DEBUG("AFSM Thread", "Gotten 2 (stop).\n");

		/* Create a new thread, and wait until it finishes. */
		pthread_create(&thr_stp, NULL, (void *) afsm_stop, info);
		pthread_join(thr_stp, NULL);
		break;
	}

	DEBUG("AFSM Thread", "Exiting multi-threaded AFSM.\n");
}

/* Auto Fan Speed Management */
static void
afsm(FL_OBJECT *obj, long data)
{
	DEBUG("AFSM", "Building AFSM Form via XForms.\n");

	/* Create new XForms objects */
	FL_FORM *afsm_form;
	FL_OBJECT *tmp, *fan, *sec, *btn;

	DEBUG("AFSM", "Adding Objects to AFSM Form.\n");

	afsm_form = fl_bgn_form(FL_UP_BOX, 320, 330);
	current_form = afsm_form;

	/* Add text to form center */
	fl_add_box(FL_UP_BOX, 160, 40, 0, 0, "Auto Fan Speed Management");
	
	/* Add sliders and handlers */
	/* CPU temperature threshold */
	tmp = fl_add_valslider(FL_HOR_SLIDER, 20, 70, 280, 30, "CPU temp threshold");
	set_valslider(tmp, MIN_TEMP, MAX_TEMP, 0, DEF_TEMP, FL_ALIGN_TOP, set_afsm_val, 0);

	/* Fan speed */
	fan = fl_add_valslider(FL_HOR_SLIDER, 20, 130, 130, 30, "Fan speed");
	set_valslider(fan, 0, 8, 0, 7, FL_ALIGN_TOP, set_speed, 1);

	/* Interval in seconds */
	sec = fl_add_valslider(FL_HOR_SLIDER, 170, 130, 130, 30, "Interval (in seconds)");
	set_valslider(sec, 0, 10, 0, 5, FL_ALIGN_TOP, set_afsm_secs, 0);

	/* Start monitoring CPU temp */
	btn = fl_add_button(FL_NORMAL_BUTTON, 20, 170, 280, 40, "Apply & Run");
	fl_set_object_callback(btn, afsm_thr, 0);

	/* Button to stop */
	btn = fl_add_button(FL_NORMAL_BUTTON, 20, 220, 280, 40, "Set fan speed to auto");
	fl_set_object_callback(btn, afsm_thr, 1);

	/* Button to close AFSM Form */
	btn = fl_add_button(FL_NORMAL_BUTTON, 20, 270, 280, 40, "Close");
	fl_set_object_callback(btn, afsm_thr, 2);

	current_form = afsm_form;

	fl_end_form();

	DEBUG("AFSM", "Done Adding Objects to Exit Form.\n");

	/* Show form */
	fl_show_form(afsm_form, FL_PLACE_CENTER, FL_TRANSIENT,
			"Auto Fan Speed Management");

	fl_do_forms();

	DEBUG("AFSM", "Showing AFSM Form.\n");

	current_form = afsm_form;

	fl_free_form(afsm_form);

	DEBUG("AFSM", "AFSM Form destroyed.\n");
}

int
main(int argc, char **argv)
{
	/* Before starting XForms, try to parse options */
	int ind = 0;

	while ((ind = getopt(argc, argv, ":dihp:v")) != 1) {
		switch (ind) {
		/* Enable debugging */
		case 'd':
			debug = 1;
			break;

		/* Print info */
		case 'i':
			printf("%s\n", INFO_STR);
			return 0;
			break;

		/* Show help */
		case '?':
		case 'h':
			printf("%s by Salonia Matteo. Report any bugs to %s.\n"
				"Support this project: %s\n"
				"\n"
				"Available flags:\n"
				"-d		Enable debugging.\n"
				"-i		Print information and exit, without opening XForms.\n"
				"-h		Print this help and exit, without opening XForms.\n"
				"-p path\t	Which directory contains the apply.sh script.\n"
				"-v		Be more verbose. (writes info to the console)\n",
				PACKAGE_STRING, PACKAGE_BUGREPORT, PACKAGE_DONATE);
			return 0;
			break;

		/* Use custom path */
		case 'p':
			strncpy(user_path, optarg, PATH_MAX);
			using_custom_path = 1;
			break;

		/* Be more verbose */
		case 'v':
			verbose = 1;
			break;
		}

		if (ind <= 0)
			break;
	}

	DEBUG("Main", "Finished parsing options, starting XForms.\n");

	FL_FORM *form;	/* Create new form */
	FL_OBJECT *obj;	/* Create new object */

	/* Initialise XForms using argc & argv */
	fl_initialize(&argc, argv, 0, 0, 0);

	/* Create a new form */
	form = fl_bgn_form(FL_UP_BOX, 250, 300);
	current_form = form;

	DEBUG("Main", "Building Main Form via XForms.\n");

	/* Add a slider to select values between 0 (auto) and 8 (max) */
	obj = fl_add_valslider(FL_HOR_SLIDER, 10, 25, 230, 40, "Fan speed (0 = auto, 8 = max)");
	set_valslider(obj, 0, 8, 0, 0, FL_ALIGN_TOP, set_speed, 0);
	
	/* Align label to left */
	fl_set_object_lalign(obj, FL_ALIGN_LEFT_TOP);

	/* Add an "Apply" button, that applies current speed */
	obj = fl_add_button(FL_NORMAL_BUTTON, 10, 80, 230, 50, "Apply");
	fl_set_object_callback(obj, apply_speed, 0);

	/* Add a button to use Auto Fan Speed Management */
	obj = fl_add_button(FL_NORMAL_BUTTON, 10, 140, 230, 50, "Auto Fan Speed Management");
	fl_set_object_callback(obj, afsm, 0);

	/* Add an "Info" button, that shows app info */
	obj = fl_add_button(FL_NORMAL_BUTTON, 10, 200, 230, 40, "Info");
	fl_set_object_callback(obj, showinfo, 0);

	/* Add a "Quit" button, that ends XForms */
	obj = fl_add_button(FL_NORMAL_BUTTON, 10, 250, 230, 40, "Quit");
	fl_set_object_callback(obj, askexit, 0);

	DEBUG("Main", "Finished building Main Form.\n");

	fl_end_form();

	DEBUG("Main", "Showing Main Form.\n");

	/* Show the form, with title "Greet" */
	fl_show_form(form, FL_PLACE_CENTER, FL_FULLBORDER, "TP-XFan by Salonia Matteo");
	fl_do_forms();

	current_form = form;

	DEBUG("Main", "Exiting Main Form, Destroying all Forms.\n");

	fl_deactivate_all_forms();

	DEBUG("Main", "Destroyed all forms.\n");

	return 0;
}
